### ISSUE TITLE
OSSL_PARAM documentation

### ISSUE THREAD

**Issue by kroeckx** [openssl/openssl#11162]
Is OSSL_PARAM_int.pod, it has this as example:
```
          if ((p = OSSL_PARAM_locate(params, "foo")) == NULL)
               OSSL_PARAM_set_utf8_ptr(p, "foo value");
```

Should that be `!= NULL`?

**Comment by kroeckx** [comment]
But even with that change, it's not making any sense to me. What is the point of the foo variable, if the responder is just going to override it, why is the call not `OSSL_PARAM_utf8_ptr("foo", NULL, 0)`?

**Comment by levitte** [comment]
It should! But also, there should be a check of OSSL_PARAM_set_utf8_ptr(), so I assume it was really meant to be something like this:

``` C
          if ((p = OSSL_PARAM_locate(params, "foo")) == NULL
              || !OSSL_PARAM_set_utf8_ptr(p, "foo value"))
              /* error */
```

**Comment by kroeckx** [comment]
The return value of OSSL_PARAM_set_utf8_ptr doesn't seem to be documented

**Comment by kroeckx** [comment]
Oh, it's `All other functions return 1 on success and 0 on failure`

**Comment by kroeckx** [comment]
I guess I misunderstood how the *_ptr() works.

**Comment by paulidale** [comment]
I think everybody (except perhaps @levitte) has misunderstood or misused this particular function.

I've been tempted on a number of occasions to add a line to the documentation: _This is (almost certainly) not the function you are looking for_.

**Comment by richsalz** [comment]
Less flippantly, it seems that more docs for "*_ptr()" seems called for.

**Comment by paulidale** [comment]
A case could be made for removing them entirely.  Currently, they appear to be used for the name, version and buildinfo strings only.

**Comment by levitte** [comment]
We probably underuse them. I haven't looked very closely, but I suspect we do more copying of constants than we know...

**Comment by kroeckx** [comment]
My problem with the OSSL_PARAM_utf8_ptr() macro is that it does &(addr). It would at least be clearer what is going on if the macro didn't take the address of the parameter you pass.

**Comment by paulidale** [comment]
The entire point of the two _PTR param types is to pass around pointers using their address rather than what they point to.  They save a `memcpy(3)` by fiddling with the pointers directly.

**Comment by kroeckx** [comment]
My problem is that you now have to do:
char *foo;
OSSL_PARAM_utf8_ptr(foo);

And my first reaction is, this can't work.

But since OSSL_PARAM_utf8_ptr is a define, it can turn it into:
OSSL_PARAM_DEFN(&foo);

I think we should instead write:
OSSL_PARAM_utf8_ptr(&foo);

It's much cleared about what that "function" does.

I also think none of this should actually be defines, but real
functions, so that we can have some type safety. The struct should
then also not be in the public header, and only the functions
should be used.

**Comment by t8m** [comment]
I am afraid of the imposed overhead if all these are functions. The OSSL_PARAM is a crucial part of the new APIs.

**Comment by levitte** [comment]
> The struct should then also not be in the public header, and only the functions should be used.

No. We don't want to have to heap allocate it all over the place.

**Comment by paulidale** [comment]
@t8m the overhead is designed to not be terrible.  Yes, they are pervasive but they shouldn't dominate any calculation.  For example, any public key algorithm will make the overhead effectively nothing.  Likewise, symmetric crypto might show an impact for small sizes, but again it amortises as the size increases (plus the overhead is lower).  I think some overhead measurements will be enlightening as to this assumption's correctness.

@levitte, @kroeckx if we go to opaque structures and functions, the PARAM builder might be a better starting point.  It's easier to use but does usually involve an extra copy.

**Comment by t8m** [comment]
@paulidale I understand the overhead of the current implementation is quite minimal. What I would not like to see is raising it by forcing allocations and function calls everywhere.

**Comment by levitte** [comment]
Another reason to not make OSSL_PARAM opaque is to make it shareable between libcrypto and providers. This is by design.

**Comment by levitte** [comment]
The original issue has a fix in #13426

**Comment by levitte** [comment]
So the original issue is now fixed.  As for the rest that's been said here, perhaps raise another issue or two?
